packages/iov-ethereum/src/serialization.ts

```diff
@@ -432,4 +340,387 @@ export class Serialization {
+      const erc20TransferCall = Serialization.buildErc20TransferCall(unsigned);
+
+      return Serialization.serializeGenericTransaction(
+        nonce,
+        gasPriceHex,
+        gasLimitHex,
+        erc20Token.contractAddress,
+        ZERO_AMOUNT, // ETH value
+        erc20TransferCall,
+        chainIdHex,
+      );
+    } else {
+      // native ETH send
+      const data = toUtf8(unsigned.memo || "");
+
+      return Serialization.serializeGenericTransaction(
+        nonce,
+        gasPriceHex,
+        gasLimitHex,
+        unsigned.recipient,
+        unsigned.amount.quantity,
+        data,
+        chainIdHex,
+      );
+    }
+  }
+
+  private static serializeUnsignedSwapOfferTransaction(
+    unsigned: SwapOfferTransaction,
+    nonce: Nonce,
+    erc20Tokens: ReadonlyMap<TokenTicker, Erc20Options>,
+    chainIdHex: string,
+    gasPriceHex: string,
+    gasLimitHex: string,
+    atomicSwapEtherContractAddress?: Address,
+    atomicSwapErc20ContractAddress?: Address,
+  ): Uint8Array {
+    Serialization.checkSwapId(unsigned);
+    Serialization.checkHash(unsigned);
+    Serialization.checkRecipientAddress(unsigned);
+    Serialization.checkMemoNotPresent(unsigned);
+    Serialization.checkAtomicSwapContractAddress(
+      atomicSwapEtherContractAddress,
+      atomicSwapErc20ContractAddress,
+    );
+    if (!isBlockHeightTimeout(unsigned.timeout)) {
+      throw new Error("Timeout must be specified as a block height");
+    }
+
+    if (atomicSwapEtherContractAddress) {
+      // native ETH swap
+      Serialization.checkEtherAmount(unsigned);
+
+      const atomicSwapOfferCall = Serialization.buildAtomicSwapOfferEtherCall(unsigned);
+
+      return Serialization.serializeGenericTransaction(
+        nonce,
+        gasPriceHex,
+        gasLimitHex,
+        atomicSwapEtherContractAddress,
+        unsigned.amounts[0].quantity,
+        atomicSwapOfferCall,
+        chainIdHex,
+      );
+    } else {
+      // ERC20 swap
+      Serialization.checkErc20Amount(unsigned, erc20Tokens);
+
+      const erc20Token = Serialization.getErc20Token(unsigned, erc20Tokens);
+      const atomicSwapOfferCall = Serialization.buildAtomicSwapOfferErc20Call(
+        unsigned,
+        erc20Token.contractAddress,
+      );
+
+      return Serialization.serializeGenericTransaction(
+        nonce,
+        gasPriceHex,
+        gasLimitHex,
+        atomicSwapErc20ContractAddress!,
+        ZERO_AMOUNT,
+        atomicSwapOfferCall,
+        chainIdHex,
+      );
+    }
+  }
+
+  private static serializeUnsignedSwapClaimTransaction(
+    unsigned: SwapClaimTransaction,
+    nonce: Nonce,
+    chainIdHex: string,
+    gasPriceHex: string,
+    gasLimitHex: string,
+    atomicSwapEtherContractAddress?: Address,
+    atomicSwapErc20ContractAddress?: Address,
+  ): Uint8Array {
+    Serialization.checkSwapId(unsigned);
+    Serialization.checkPreimage(unsigned);
+    Serialization.checkAtomicSwapContractAddress(
```

Comment:
When this checker returns the one contract address in the success case, you don't need to force unwrap (`!`) below and get better type safety.